File: Sheets/src/neetcode/practice/Binary_Search/_3_kokoEatingBananas.java
package neetcode.practice.Binary_Search;

import java.util.Arrays;

public class _3_kokoEatingBananas {
  static int totalTime(int[] piles, int rate, int n) {
    int ans = 0;
    for (int i = 0; i < n; i++) ans += Math.ceil((double) piles[i] / rate);
    return ans;
  }

  public static int minEatingSpeed(int[] piles, int h) {
    int l = 1;
    int r = Arrays.stream(piles).max().getAsInt();
    int ans = r, n = piles.length;

    while (l <= r) {
      int m = (l + r) / 2;
      if (totalTime(piles, m, n) <= h) {
        ans = m;
        r = m - 1;
      } else l = m + 1;
    }

    return ans;
  }

  public static void main(String[] args) {
    int[] piles = {25, 10, 23, 4};
    System.out.println(minEatingSpeed(piles, 4));
  }
}

File: Sheets/src/neetcode/practice/Binary_Search/_3_KokoEatingBananas.cc
#include <algorithm>
#include <cmath>
#include <vector>

class Solution
{
public:
  int minEatingSpeed(std::vector<int> &piles, int h)
  {
    int L{0}, R{*std::max_element(piles.begin(), piles.end()) + 1};

    auto predicate = [&piles, h](int rate) -> bool {
      int total_time{};
      for (const int pile : piles)
        total_time += (pile + rate - 1) / rate;

      return total_time <= h;
    };

    while (R - L > 1)
    {
      int M = L + ((R - L) >> 1);
      if (predicate(M))
        R = M;
      else
        L = M;
    }

    return R;
  }
};
